cli/emit_cache.rs

```diff
@@ -0,0 +1,324 @@
+use std::cell::RefCell;
+use std::collections::HashMap;
+use std::path::Path;
+
+use deno_ast::ModuleSpecifier;
+use deno_core::error::AnyError;
+use rusqlite::params;
+use rusqlite::Connection;
+use serde::Deserialize;
+use serde::Serialize;
+
+// Bump these to invalidate the caches on a row by row basis
+const EMIT_DATA_VERSION: usize = 1;
+const TS_BUILD_INFO_VERSION: usize = 1;
+
+// todo: remove serialize, deserialize BEFORE PR
+/// Emit cache for a single file.
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct EmitCacheData {
+  pub source_hash: String,
+  pub text: String,
+  pub map: Option<String>,
+  pub declaration: Option<String>,
+}
+
+pub trait EmitCache {
+  /// Gets the emit data from the cache.
+  fn get_emit_data(&self, specifier: &ModuleSpecifier)
+    -> Option<EmitCacheData>;
+  /// Sets the emit data in the cache.
+  fn set_emit_data(
+    &self,
+    specifier: &ModuleSpecifier,
+    data: &EmitCacheData,
+  ) -> Result<(), AnyError>;
+
+  /// Gets the .tsbuildinfo file from the cache.
+  fn get_tsbuildinfo(&self, specifier: &ModuleSpecifier) -> Option<String>;
+  /// Sets the .tsbuildinfo file in the cache.
+  fn set_tsbuildinfo(
+    &self,
+    specifier: &ModuleSpecifier,
+    text: &str,
+  ) -> Result<(), AnyError>;
+}
+
+#[derive(Default)]
+pub struct MemoryEmitCache {
+  build_infos: RefCell<HashMap<ModuleSpecifier, String>>,
+  emits: RefCell<HashMap<ModuleSpecifier, EmitCacheData>>,
+}
+
+impl EmitCache for MemoryEmitCache {
+  fn get_emit_data(
+    &self,
+    specifier: &ModuleSpecifier,
+  ) -> Option<EmitCacheData> {
+    self.emits.borrow().get(specifier).cloned()
+  }
+
+  fn set_emit_data(
+    &self,
+    specifier: &ModuleSpecifier,
+    data: &EmitCacheData,
+  ) -> Result<(), AnyError> {
+    self
+      .emits
+      .borrow_mut()
+      .insert(specifier.clone(), data.clone());
+    Ok(())
+  }
+
+  fn get_tsbuildinfo(&self, specifier: &ModuleSpecifier) -> Option<String> {
+    self.build_infos.borrow().get(specifier).cloned()
+  }
+
+  fn set_tsbuildinfo(
+    &self,
+    specifier: &ModuleSpecifier,
+    text: &str,
+  ) -> Result<(), AnyError> {
+    self
+      .build_infos
+      .borrow_mut()
+      .insert(specifier.clone(), text.to_string());
+    Ok(())
+  }
+}
+
+pub struct SqliteEmitCache {
+  conn: Connection,
+  /// these stored here for mutating for testing purposes
+  emit_data_version: usize,
+  ts_build_info_version: usize,
+}
+
+impl SqliteEmitCache {
+  pub fn new(db_file_path: &Path) -> Result<Self, AnyError> {
+    let conn = Connection::open(db_file_path)?;
+    Self::from_connection(conn)
+  }
```

Comment:
Consider using these pragmas, they can significantly speed things up. https://github.com/denoland/deno/blob/56d0ca7f52d9bff7326bb2b9406ea1e0282a5fbe/ext/webstorage/lib.rs#L69-L81

Reply:
Thanks for letting me know! I added them 🙂